File: php/src/Resources/TracksResource.php
<?php

declare(strict_types=1);

namespace JewelMusic\Resources;

/**
 * Tracks resource for track upload, metadata management, and organization
 */
class TracksResource extends BaseResource
{
    /**
     * Upload a track with metadata
     *
     * @param string|resource $file Audio file to upload
     * @param string $filename Original filename
     * @param array $metadata Track metadata
     *                       - title: Track title (required)
     *                       - artist: Artist name (required)
     *                       - album: Album name
     *                       - genre: Music genre
     *                       - releaseDate: Release date
     *                       - duration: Track duration
     *                       - explicit: Explicit content flag
     *                       - isrc: ISRC code
     * @param array $options Upload options
     *                      - chunkSize: Size for chunked upload
     *                      - quality: Audio quality preference
     * @return array Uploaded track data
     */
    public function upload($file, string $filename, array $metadata, array $options = []): array
    {
        $this->validateRequired($metadata, ['title', 'artist']);
        
        $uploadMetadata = [];
        foreach ($metadata as $key => $value) {
            if (is_bool($value)) {
                $uploadMetadata[$key] = $value ? 'true' : 'false';
            } else {
                $uploadMetadata[$key] = (string)$value;
            }
        }
        
        // Add upload options to metadata
        foreach ($options as $key => $value) {
            if ($value !== null && $value !== '') {
                $uploadMetadata[$key] = (string)$value;
            }
        }
        
        $response = $this->httpClient->uploadFile('/tracks/upload', $file, $filename, $uploadMetadata);
        return $this->extractData($response);
    }

    /**
     * Upload track with chunked upload for large files
     *
     * @param string|resource $file Audio file to upload
     * @param string $filename Original filename
     * @param array $metadata Track metadata
     * @param int $chunkSize Chunk size in bytes (default: 8MB)
     * @return array Uploaded track data
     */
    public function uploadChunked($file, string $filename, array $metadata, int $chunkSize = 8388608): array
    {
        $this->validateRequired($metadata, ['title', 'artist']);
        
        $uploadMetadata = [];
        foreach ($metadata as $key => $value) {
            if (is_bool($value)) {
                $uploadMetadata[$key] = $value ? 'true' : 'false';
            } else {
                $uploadMetadata[$key] = (string)$value;
            }
        }
        
        $response = $this->httpClient->uploadFileChunked('/tracks/upload', $file, $filename, $uploadMetadata, $chunkSize);
        return $this->extractData($response);
    }

    /**
     * Get track by ID
     *
     * @param string $trackId Track ID
     * @return array Track data
     */
    public function get(string $trackId): array
    {
        $response = $this->httpClient->get("/tracks/{$trackId}");
        return $this->extractData($response);
    }

    /**
     * Update track metadata
     *
     * @param string $trackId Track ID
     * @param array $metadata Updated metadata
     * @return array Updated track data
     */
    public function update(string $trackId, array $metadata): array
    {
        $this->validateRequired(['trackId' => $trackId], ['trackId']);
        
        $response = $this->httpClient->put("/tracks/{$trackId}", $metadata);
        return $this->extractData($response);
    }

    /**
     * Delete a track
     *
     * @param string $trackId Track ID
     * @return array Deletion confirmation
     */
    public function delete(string $trackId): array
    {
        $this->validateRequired(['trackId' => $trackId], ['trackId']);
        
        $response = $this->httpClient->delete("/tracks/{$trackId}");
        return $this->extractData($response);
    }

    /**
     * Upload artwork for a track
     *
     * @param string $trackId Track ID
     * @param string|resource $artworkFile Artwork image file
     * @param string $filename Artwork filename
     * @return array Artwork upload result
     */
    public function uploadArtwork(string $trackId, $artworkFile, string $filename): array
    {
        $this->validateRequired([
            'trackId' => $trackId,
            'artworkFile' => $artworkFile,
            'filename' => $filename
        ], ['trackId', 'artworkFile', 'filename']);
        
        $response = $this->httpClient->uploadFile("/tracks/{$trackId}/artwork", $artworkFile, $filename);
        return $this->extractData($response);
    }

    /**
     * Batch update metadata for multiple tracks
     *
     * @param array $updates Array of track updates
     *                      Each item should contain 'id' and 'metadata' keys
     * @return array Batch update results
     */
    public function batchUpdateMetadata(array $updates): array
    {
        foreach ($updates as $update) {
            $this->validateRequired($update, ['id', 'metadata']);
        }
        
        $data = ['updates' => $updates];
        $response = $this->httpClient->post('/tracks/batch/metadata', $data);
        return $this->extractData($response);
    }

    /**
     * Queue tracks for batch processing
     *
     * @param array $trackIds Array of track IDs
     * @param array $operations Processing operations to perform
     * @param array $options Batch processing options
     *                      - priority: Processing priority
     *                      - notify: Send notification when complete
     * @return array Batch processing job data
     */
    public function batchProcess(array $trackIds, array $operations, array $options = []): array
    {
        $this->validateRequired(['trackIds' => $trackIds, 'operations' => $operations], ['trackIds', 'operations']);
        
        $data = array_merge([
            'trackIds' => $trackIds,
            'operations' => $operations
        ], $this->filterNullValues($options));
        
        $response = $this->httpClient->post('/tracks/batch/process', $data);
        return $this->extractData($response);
    }

    /**
     * Get track processing status
     *
     * @param string $trackId Track ID
     * @return array Processing status information
     */
    public function getProcessingStatus(string $trackId): array
    {
        $response = $this->httpClient->get("/tracks/{$trackId}/processing-status");
        return $this->extractData($response);
    }

    /**
     * Generate waveform visualization for a track
     *
     * @param string $trackId Track ID
     * @param array $options Waveform generation options
     *                      - width: Image width in pixels
     *                      - height: Image height in pixels
     *                      - colors: Color scheme for waveform
     *                      - format: Output format (png, svg, json)
     *                      - samples: Number of sample points
     * @return array Waveform data or image URL
     */
    public function generateWaveform(string $trackId, array $options = []): array
    {
        $response = $this->httpClient->post("/tracks/{$trackId}/waveform", $this->filterNullValues($options));
        return $this->extractData($response);
    }

    /**
     * Get download URL for a track
     *
     * @param string $trackId Track ID
     * @param string $format Audio format (mp3, wav, flac, etc.)
     * @param string $quality Audio quality (high, medium, low)
     * @return array Download URL and metadata
     */
    public function getDownloadUrl(string $trackId, string $format = 'mp3', string $quality = 'high'): array
    {
        $params = [
            'format' => $format,
            'quality' => $quality
        ];
        
        $response = $this->httpClient->get("/tracks/{$trackId}/download", $params);
        return $this->extractData($response);
    }

    /**
     * Find similar tracks based on content analysis
     *
     * @param string $referenceTrackId Reference track ID
     * @param array $options Similarity search options
     *                      - limit: Maximum number of results
     *                      - minSimilarity: Minimum similarity threshold
     *                      - sameArtist: Include tracks from same artist
     *                      - sameGenre: Include tracks from same genre
     *                      - features: Specific features to compare
     * @return array Similar tracks
     */
    public function findSimilar(string $referenceTrackId, array $options = []): array
    {
        $params = $this->buildParams([], $options);
        $response = $this->httpClient->get("/tracks/{$referenceTrackId}/similar", $params);
        return $this->extractData($response);
    }

    /**
     * Get track audio features and characteristics
     *
     * @param string $trackId Track ID
     * @param array $features Specific features to retrieve
     *                       - tempo: BPM and rhythmic features
     *                       - key: Musical key and scale
     *                       - energy: Energy and intensity levels
     *                       - mood: Emotional characteristics
     * @return array Audio features
     */
    public function getAudioFeatures(string $trackId, array $features = []): array
    {
        $params = $this->buildParams([], ['features' => $features]);
        $response = $this->httpClient->get("/tracks/{$trackId}/features", $params);
        return $this->extractData($response);
    }

    /**
     * Create playlist from tracks
     *
     * @param array $trackIds Array of track IDs
     * @param array $playlistData Playlist metadata
     *                           - name: Playlist name
     *                           - description: Playlist description
     *                           - public: Public visibility
     * @return array Created playlist data
     */
    public function createPlaylist(array $trackIds, array $playlistData): array
    {
        $this->validateRequired($playlistData, ['name']);
        
        $data = array_merge(['trackIds' => $trackIds], $playlistData);
        $response = $this->httpClient->post('/tracks/playlists', $data);
        return $this->extractData($response);
    }

    /**
     * Tag tracks with custom labels
     *
     * @param string $trackId Track ID
     * @param array $tags Array of tags to add
     * @return array Updated track with tags
     */
    public function addTags(string $trackId, array $tags): array
    {
        $data = ['tags' => $tags];
        $response = $this->httpClient->post("/tracks/{$trackId}/tags", $data);
        return $this->extractData($response);
    }

    /**
     * Remove tags from track
     *
     * @param string $trackId Track ID
     * @param array $tags Array of tags to remove
     * @return array Updated track
     */
    public function removeTags(string $trackId, array $tags): array
    {
        $data = ['tags' => $tags];
        $response = $this->httpClient->delete("/tracks/{$trackId}/tags", $data);
        return $this->extractData($response);
    }

    /**
     * List tracks with filtering and pagination
     *
     * @param int $page Page number
     * @param int $perPage Items per page
     * @param array $filters Track filters
     *                      - status: Processing status
     *                      - genre: Genre filter
     *                      - artist: Artist name filter
     *                      - album: Album filter
     *                      - uploadedAfter: Date filter
     *                      - uploadedBefore: Date filter
     *                      - durationMin: Minimum duration
     *                      - durationMax: Maximum duration
     *                      - search: Text search in metadata
     * @return array Tracks list
     */
    public function list(int $page = 1, int $perPage = 20, array $filters = []): array
    {
        $params = $this->buildParams([
            'page' => (string)$page,
            'perPage' => (string)$perPage
        ], $filters);
        
        $response = $this->httpClient->get('/tracks', $params);
        return $this->extractData($response);
    }
}

File: ruby/lib/jewelmusic/resources/tracks_resource.rb
# frozen_string_literal: true

module JewelMusic
  module Resources
    ##
    # Tracks resource for track upload, metadata management, and organization
    #
    class TracksResource < BaseResource
      ##
      # Upload a track with metadata
      #
      # @param file [String, File, IO] Audio file to upload
      # @param filename [String] Original filename
      # @param metadata [Hash] Track metadata
      # @option metadata [String] :title Track title (required)
      # @option metadata [String] :artist Artist name (required)
      # @option metadata [String] :album Album name
      # @option metadata [String] :genre Music genre
      # @option metadata [String] :release_date Release date
      # @option metadata [Integer] :duration Track duration
      # @option metadata [Boolean] :explicit Explicit content flag
      # @option metadata [String] :isrc ISRC code
      # @param options [Hash] Upload options
      # @option options [Integer] :chunk_size Size for chunked upload
      # @option options [String] :quality Audio quality preference
      # @return [Hash] Uploaded track data
      #
      def upload(file, filename, metadata, options = {})
        validate_required(metadata, [:title, :artist])
        
        prepared_file = prepare_file(file)
        upload_metadata = prepare_upload_metadata(metadata.merge(options))
        
        response = http_client.upload_file('/tracks/upload', prepared_file, filename, upload_metadata)
        extract_data(response)
      end

      ##
      # Upload track with chunked upload for large files
      #
      # @param file [String, File, IO] Audio file to upload
      # @param filename [String] Original filename
      # @param metadata [Hash] Track metadata
      # @param chunk_size [Integer] Chunk size in bytes (default: 8MB)
      # @return [Hash] Uploaded track data
      #
      def upload_chunked(file, filename, metadata, chunk_size = 8_388_608)
        validate_required(metadata, [:title, :artist])
        
        prepared_file = prepare_file(file)
        upload_metadata = prepare_upload_metadata(metadata)
        
        response = http_client.upload_file_chunked('/tracks/upload', prepared_file, filename, upload_metadata, chunk_size)
        extract_data(response)
      end

      ##
      # Get track by ID
      #
      # @param track_id [String] Track ID
      # @return [Hash] Track data
      #
      def get(track_id)
        validate_required({ track_id: track_id }, [:track_id])
        response = http_client.get("/tracks/#{track_id}")
        extract_data(response)
      end

      ##
      # Update track metadata
      #
      # @param track_id [String] Track ID
      # @param metadata [Hash] Updated metadata
      # @return [Hash] Updated track data
      #
      def update(track_id, metadata)
        validate_required({ track_id: track_id }, [:track_id])
        
        response = http_client.put("/tracks/#{track_id}", metadata)
        extract_data(response)
      end

      ##
      # Delete a track
      #
      # @param track_id [String] Track ID
      # @return [Hash] Deletion confirmation
      #
      def delete(track_id)
        validate_required({ track_id: track_id }, [:track_id])
        
        response = http_client.delete("/tracks/#{track_id}")
        extract_data(response)
      end

      ##
      # Upload artwork for a track
      #
      # @param track_id [String] Track ID
      # @param artwork_file [String, File, IO] Artwork image file
      # @param filename [String] Artwork filename
      # @return [Hash] Artwork upload result
      #
      def upload_artwork(track_id, artwork_file, filename)
        validate_required({
          track_id: track_id,
          artwork_file: artwork_file,
          filename: filename
        }, [:track_id, :artwork_file, :filename])
        
        prepared_file = prepare_file(artwork_file)
        response = http_client.upload_file("/tracks/#{track_id}/artwork", prepared_file, filename)
        extract_data(response)
      end

      ##
      # Batch update metadata for multiple tracks
      #
      # @param updates [Array<Hash>] Array of track updates
      #   Each item should contain :id and :metadata keys
      # @return [Hash] Batch update results
      #
      def batch_update_metadata(updates)
        updates.each do |update|
          validate_required(update, [:id, :metadata])
        end
        
        data = { updates: updates }
        response = http_client.post('/tracks/batch/metadata', data)
        extract_data(response)
      end

      ##
      # Queue tracks for batch processing
      #
      # @param track_ids [Array<String>] Array of track IDs
      # @param operations [Array<String>] Processing operations to perform
      # @param options [Hash] Batch processing options
      # @option options [String] :priority Processing priority
      # @option options [Boolean] :notify Send notification when complete
      # @return [Hash] Batch processing job data
      #
      def batch_process(track_ids, operations, options = {})
        validate_required({ track_ids: track_ids, operations: operations }, [:track_ids, :operations])
        
        data = {
          track_ids: track_ids,
          operations: operations
        }.merge(filter_nil_values(options))
        
        response = http_client.post('/tracks/batch/process', data)
        extract_data(response)
      end

      ##
      # Get track processing status
      #
      # @param track_id [String] Track ID
      # @return [Hash] Processing status information
      #
      def get_processing_status(track_id)
        validate_required({ track_id: track_id }, [:track_id])
        response = http_client.get("/tracks/#{track_id}/processing-status")
        extract_data(response)
      end

      ##
      # Generate waveform visualization for a track
      #
      # @param track_id [String] Track ID
      # @param options [Hash] Waveform generation options
      # @option options [Integer] :width Image width in pixels
      # @option options [Integer] :height Image height in pixels
      # @option options [Array<String>] :colors Color scheme for waveform
      # @option options [String] :format Output format (png, svg, json)
      # @option options [Integer] :samples Number of sample points
      # @return [Hash] Waveform data or image URL
      #
      def generate_waveform(track_id, options = {})
        validate_required({ track_id: track_id }, [:track_id])
        response = http_client.post("/tracks/#{track_id}/waveform", filter_nil_values(options))
        extract_data(response)
      end

      ##
      # Get download URL for a track
      #
      # @param track_id [String] Track ID
      # @param format [String] Audio format (mp3, wav, flac, etc.)
      # @param quality [String] Audio quality (high, medium, low)
      # @return [Hash] Download URL and metadata
      #
      def get_download_url(track_id, format = 'mp3', quality = 'high')
        validate_required({ track_id: track_id }, [:track_id])
        
        params = {
          format: format,
          quality: quality
        }
        
        response = http_client.get("/tracks/#{track_id}/download", params)
        extract_data(response)
      end

      ##
      # Find similar tracks based on content analysis
      #
      # @param reference_track_id [String] Reference track ID
      # @param options [Hash] Similarity search options
      # @option options [Integer] :limit Maximum number of results
      # @option options [Float] :min_similarity Minimum similarity threshold
      # @option options [Boolean] :same_artist Include tracks from same artist
      # @option options [Boolean] :same_genre Include tracks from same genre
      # @option options [Array<String>] :features Specific features to compare
      # @return [Hash] Similar tracks
      #
      def find_similar(reference_track_id, options = {})
        validate_required({ reference_track_id: reference_track_id }, [:reference_track_id])
        params = build_params({}, options)
        response = http_client.get("/tracks/#{reference_track_id}/similar", params)
        extract_data(response)
      end

      ##
      # Get track audio features and characteristics
      #
      # @param track_id [String] Track ID
      # @param features [Array<String>] Specific features to retrieve
      #   Options: tempo, key, energy, mood
      # @return [Hash] Audio features
      #
      def get_audio_features(track_id, features = [])
        validate_required({ track_id: track_id }, [:track_id])
        params = build_params({}, { features: features })
        response = http_client.get("/tracks/#{track_id}/features", params)
        extract_data(response)
      end

      ##
      # Create playlist from tracks
      #
      # @param track_ids [Array<String>] Array of track IDs
      # @param playlist_data [Hash] Playlist metadata
      # @option playlist_data [String] :name Playlist name (required)
      # @option playlist_data [String] :description Playlist description
      # @option playlist_data [Boolean] :public Public visibility
      # @return [Hash] Created playlist data
      #
      def create_playlist(track_ids, playlist_data)
        validate_required(playlist_data, [:name])
        validate_required({ track_ids: track_ids }, [:track_ids])
        
        data = { track_ids: track_ids }.merge(playlist_data)
        response = http_client.post('/tracks/playlists', data)
        extract_data(response)
      end

      ##
      # Tag tracks with custom labels
      #
      # @param track_id [String] Track ID
      # @param tags [Array<String>] Array of tags to add
      # @return [Hash] Updated track with tags
      #
      def add_tags(track_id, tags)
        validate_required({ track_id: track_id, tags: tags }, [:track_id, :tags])
        data = { tags: tags }
        response = http_client.post("/tracks/#{track_id}/tags", data)
        extract_data(response)
      end

      ##
      # Remove tags from track
      #
      # @param track_id [String] Track ID
      # @param tags [Array<String>] Array of tags to remove
      # @return [Hash] Updated track
      #
      def remove_tags(track_id, tags)
        validate_required({ track_id: track_id, tags: tags }, [:track_id, :tags])
        data = { tags: tags }
        response = http_client.delete("/tracks/#{track_id}/tags", data)
        extract_data(response)
      end

      ##
      # List tracks with filtering and pagination
      #
      # @param page [Integer] Page number
      # @param per_page [Integer] Items per page
      # @param filters [Hash] Track filters
      # @option filters [String] :status Processing status
      # @option filters [String] :genre Genre filter
      # @option filters [String] :artist Artist name filter
      # @option filters [String] :album Album filter
      # @option filters [String] :uploaded_after Date filter
      # @option filters [String] :uploaded_before Date filter
      # @option filters [Integer] :duration_min Minimum duration
      # @option filters [Integer] :duration_max Maximum duration
      # @option filters [String] :search Text search in metadata
      # @return [Hash] Tracks list
      #
      def list(page = 1, per_page = 20, filters = {})
        params = build_params({
          page: page.to_s,
          per_page: per_page.to_s
        }, filters)
        
        response = http_client.get('/tracks', params)
        extract_data(response)
      end
    end
  end
end